### PR TITLE
Allow simplification of singleton interval.

### LINKED ISSUE #7690: Spurious pointer check violation with smt2 backend and z3 4.12.1
This program is proved correct with the SAT backend but fails with with SMT2 backend with `--pointer-check --bounds-check`:

```
[stk_push.pointer_dereference.17] line 57 dereference failure: pointer outside object bounds in stk->elems[(signed long int)stk->top]: FAILURE
```

The analysis using the SMT2 backend succeeds if we change the assumption at line 46:

``
  __CPROVER_assume(UINT8_MAX <= stk_size && stk_size <= UINT8_MAX);
``

into 

```
  __CPROVER_assume(UINT8_MAX == stk_size);
```

or into an assignment

```
stk_size = UINT8_MAX;
```

```c
#include <stdlib.h>
#include <stdint.h>

extern size_t __CPROVER_max_malloc_size;
int __builtin_clzll(unsigned long long);

#define __nof_symex_objects                                                    \
  ((size_t)(1ULL << __builtin_clzll(__CPROVER_max_malloc_size)))

typedef struct {
  size_t k;
  void **ptrs;
} smap_t;

void smap_init(smap_t *smap, size_t k) {
  *smap = (smap_t){
      .k = k, .ptrs = __CPROVER_allocate(__nof_symex_objects * sizeof(void *), 1)};
}

void *smap_get(smap_t *smap, void *ptr) {
  size_t id = __CPROVER_POINTER_OBJECT(ptr);
  void *sptr = smap->ptrs[id];
  if (!sptr) {
    sptr = __CPROVER_allocate(smap->k * __CPROVER_OBJECT_SIZE(ptr), 1);
    smap->ptrs[id] = sptr;
  }
  return sptr + smap->k * __CPROVER_POINTER_OFFSET(ptr);
}

typedef struct {
  uint8_t key;
  uint8_t value;
} stk_elem_t;

typedef struct {
  int8_t top;
  stk_elem_t *elems;
} stk_t;

size_t nondet_size_t();

// Creates a fresh borrow stack
stk_t *stk_new() {
  stk_t *stk = __CPROVER_allocate(sizeof(stk_t), 1);
  size_t stk_size = nondet_size_t();
  __CPROVER_assume(UINT8_MAX <= stk_size && stk_size <= UINT8_MAX);
  // works with
  // __CPROVER_assume(stk_size == UINT8_MAX);
  *stk = (stk_t){
      .top = 0,
      .elems = __CPROVER_allocate(sizeof(stk_elem_t) * stk_size, 1)};
  return stk;
}

void stk_push(stk_t *stk, uint8_t key, uint8_t value) {
  assert(stk->top < UINT8_MAX);
  stk->elems[stk->top] = (stk_elem_t){.key = key, .value = value};
  stk->top++;
}

stk_t *get_stk(smap_t *smap, void *ptr) {
  stk_t **stk_ptr = (stk_t **) smap_get(smap, ptr);
  if (!(*stk_ptr)) {
    *stk_ptr = stk_new();
  }
  return *stk_ptr;
}

typedef struct {
  int a;
  int b;
} my_struct_t;

int main() {
  smap_t smap;
  smap_init(&smap, sizeof(stk_t*));
  my_struct_t my_struct;
  stk_t *stk_a = get_stk(&smap, &(my_struct.a));
  stk_push(stk_a, 1, 1);
  stk_t *stk_b = get_stk(&smap, &(my_struct.b));
  assert(stk_b);
  stk_push(stk_b, 1, 1);
}
```

CBMC version: 5.79.0 (cbmc-5.79.0-49-gb34e991f30), Z3 version 4.12.1 - 64 bit
Operating system: macOS
Exact command line resulting in the issue:`cbmc --pointer-check --smt2 stack_push.c`
What behaviour did you expect:  analysis suceeds
What happened instead: analysis fails

### PR DESCRIPTION
**Pull request by NlightNFotis**
Before this change, an expression like `value >= 255 && value <= 255` would fail to be simplified to `value == 255` by the expression simplifier.

~While the simplification itself is useful, the *lack* of it was causing some misbehaviour in one of our backends (the old SMT one under `src/solvers/smt2`) because the lack of simplification resulted in less constraints being propagated to the solver, missing some assertions around the expression that was using `value` as the size of an array.~

This was originally presented as a fix for #7690 , but #7768 is a more robust fix for that, so I'm changing the presentation of this as improvement to the expression simplifier instead of a workaround for an issue with the old SMT backend.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by tautschnig** [comment]
> While the simplification itself is useful, the _lack_ of it was causing some misbehaviour in one of our backends (the old SMT one under `src/solvers/smt2`) because the lack of simplification resulted in less constraints being propagated to the solver, missing some assertions around the expression that was using `value` as the size of an array.

While the simplification is nice to have we must not rely on it in our back-ends. So we also need a back-end fix here plus a test that uses —no-simplify.

**Comment by NlightNFotis** [comment]
@tautschnig Hi Michael, agree that this is a workaround for the issue, but also believe that the simplification has value itself.

I believe that this should go in on the basis that it sidesteps an important bug #7690, and I have also noted that we may need to do a deeper dive to fix the underlying issue at the root level, but it's also worth noting that this issue was only observed on the old SMT backend (the *new* SMT backend and the SAT backend both worked before).

Given that the issue is not observable on the new SMT backend however, and that it's scheduled to (eventually?) take over, it might not be worth to have deeper dive unless the issue comes up again.

File: regression/cbmc/simplify_singleton_interval_7690/singleton_interval_in_assume_7690.c [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+} smap_t;
+
+void smap_init(smap_t *smap, size_t k) {
+  *smap = (smap_t){
+      .k = k, .ptrs = __CPROVER_allocate(__nof_symex_objects * sizeof(void *), 1)};
+}
+
+void *smap_get(smap_t *smap, void *ptr) {
+  size_t id = __CPROVER_POINTER_OBJECT(ptr);
+  void *sptr = smap->ptrs[id];
```
Comment (thomasspriggs) [inline regression/cbmc/simplify_singleton_interval_7690/singleton_interval_in_assume_7690.c]: I think this test will pass on the VS2019 job if you change this line to `char *sptr = smap->ptrs[id];`

Comment (NlightNFotis) [reply regression/cbmc/simplify_singleton_interval_7690/singleton_interval_in_assume_7690.c]: Yeah that seems to work.

File: src/util/simplify_expr_boolean.cpp [hunk 2]
```diff
@@ -137,6 +139,103 @@ simplify_exprt::resultt<> simplify_exprt::simplify_boolean(const exprt &expr)
...
+            }
+            return false;
+          }
+          return false;
+        }
+        return false;
+      };
+
+      // We need to match both operands, at the particular sequence we expect.
+      structure_matched |= match_first_operand(new_operands[0]);
```
Comment (esteffin) [inline src/util/simplify_expr_boolean.cpp]: ⛏️ The variable `structure_matched` is set roughly 50 lines above and never used, so I suggest defining it here and assigning to `match_first_operand(new_operands[0])`.

`bool structure_matched = match_first_operand(new_operands[0]);`

Or even better
`bool structure_matched = match_first_operand(new_operands[0]) && match_second_operand(new_operands[1])`.

This will also be faster as the `&&` will short-circuit if the former condition is false.

File: src/util/simplify_expr_boolean.cpp [hunk 3]
```diff
@@ -137,6 +139,103 @@ simplify_exprt::resultt<> simplify_exprt::simplify_boolean(const exprt &expr)
...
+      structure_matched &= match_second_operand(new_operands[1]);
+
+      if(structure_matched && bounds.lower == bounds.higher)
+      {
+        // Go through the expression again and convert >= operand into ==
+        for(const auto &op : new_operands)
+        {
+          if(
+            const auto ge_expr =
+              expr_try_dynamic_cast<greater_than_or_equal_exprt>(op))
```
Comment (esteffin) [inline src/util/simplify_expr_boolean.cpp]: This loop and check are redundant.

We already know that `new_operands.size() == 2` ( line 148) and that they are both `greater_than_or_equal_exprt` (lines 170 and 191), so I suggest to check only `structure_matched && bounds.lower == bounds.higher` and return `new_expr{ge_expr->lhs(), ge_expr->rhs()}`.

File: src/solvers/flattening/boolbv_quantifier.cpp [hunk 4]
```diff
@@ -106,24 +106,42 @@ get_quantifier_var_max(const exprt &var_expr, const exprt &quantifier_expr)
...
+        {
+          return to_constant_expr(y_binary.rhs());
+        }
+      }
+      else
+      {
+        // If we are here, we came across a (simplified?) expression that was
+        // not anticipated - normally this would be a bug, but if you made
+        // changes to the simplifier (as an example), you would need to add an
+        // else-if branch that handles that type above.
```
Comment (esteffin) [inline src/solvers/flattening/boolbv_quantifier.cpp]: Isn't this shouting `INVARIANT` loud?

No, then the comment should be changed to specify that if we are here we may have a non-simplified expression

File: regression/cbmc-primitives/alternating_quantifiers_6231/forall_in_exists.c [hunk 5]
```diff
@@ -8,7 +8,7 @@ int main(int argc, char **argv)
 
     __CPROVER_assert(
         __CPROVER_exists { int z; (0 < z && z < 2) &&
-            __CPROVER_forall { int o; (10 < o && o < 20) ==> o > z && z == * i }},
```
Comment (esteffin) [inline regression/cbmc-primitives/alternating_quantifiers_6231/forall_in_exists.c:11]: ⛏️ The commit message looks split between title and body...

**Review by peterschrammel — COMMENTED** [review]
> causing some misbehaviour

What precisely does "misbehaviour" mean here? 

If it returns _incorrect_ results then I agree with @tautschnig that we need to fix our code so that we get correct results even with --no-simplify.

If it results just in a performance issue of the SMT backend then I'm fine with this PR as is.

**Review by esteffin — COMMENTED** [review]
Few simplification can be done

**Comment by NlightNFotis** [comment]
@peterschrammel 

Misbehaviour in this context is an assertion failing when it wouldn't fail for the SAT or the *new* SMT backend.

I played around with the code and there does remain an issue with the old SMT2 backend and this particular assertion if the interval is not the singleton interval (as in, the interval is wider than 1 element), so there will probably need to be a fix for that issue, **but** outside of that, this PR allows for improving performance, at least in the context of the SAT backend, with the following numbers coming out of the runs on the `regression/cbmc/simplify_singleton_interval_7690/singleton_interval_in_assume_7690.c` test:

* without optimisation:

```
Running with 8 object bits, 56 offset bits (default)
Starting Bounded Model Checking
Runtime Symex: 0.0659747s
size of program expression: 291 steps
simple slicing removed 14 assignments
Generated 157 VCC(s), 49 remaining after simplification
Runtime Postprocess Equation: 3.0125e-05s
Passing problem to propositional reduction
converting SSA
Runtime Convert SSA: 0.0392838s
Running propositional reduction
Post-processing
Runtime Post-process: 0.0141315s
Solving with MiniSAT 2.2.1 with simplifier
8868 variables, 5507 clauses
SAT checker: instance is UNSATISFIABLE
Runtime Solver: 0.0464093s
Runtime decision procedure: 0.0861215s
```

* With optimisation:

```
Starting Bounded Model Checking
Runtime Symex: 0.118682s
size of program expression: 288 steps
simple slicing removed 14 assignments
Generated 157 VCC(s), 48 remaining after simplification
Runtime Postprocess Equation: 2.8459e-05s
Passing problem to propositional reduction
converting SSA
Runtime Convert SSA: 0.110274s
Running propositional reduction
Post-processing
Runtime Post-process: 0.00303258s
Solving with MiniSAT 2.2.1 with simplifier
7063 variables, 2018 clauses
SAT checker inconsistent: instance is UNSATISFIABLE
Runtime Solver: 0.003055s
Runtime decision procedure: 0.113702s
```

**Comment by peterschrammel** [comment]
Is this a general bug in the old SMT backend that non-constant dynamic object sizes lead to incorrect results?

**Comment by thomasspriggs** [comment]
> Is this a general bug in the old SMT backend that non-constant dynamic object sizes lead to incorrect results?

The old SMT2 backend does just skip defining the `object_size` if it is non-constant - https://github.com/diffblue/cbmc/blob/ce831a09294434871a91fbe94abb2aa97355fa99/src/solvers/smt2/smt2_conv.cpp#L301-L314
It does this because the size term is solely implemented as a bit vector constant.

The new back end just converts the size expression into an SMT term. As such, it works for more complex expressions - https://github.com/diffblue/cbmc/blob/ce831a09294434871a91fbe94abb2aa97355fa99/src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp#L590-L591

For the example given in https://github.com/diffblue/cbmc/issues/7690 the new SMT back end produces the same verification results as the SAT backend. As in the issue is not reproducible. So this difference between the new and old SMT backends could be the source of the bug.

**Comment by peterschrammel** [comment]
If I understand correctly this means that the old SMT backend is simply broken (and probably has always been) and the simplification hides the bug in some specific cases.
So, in this sense this PR makes the old SMT backend practically useable in more cases, but it remains broken. I'd say this is an improvement. 
Still, it would be great to fix the bug in the old backend if effort allows.

**Review by peterschrammel — APPROVED** [review]

File: src/solvers/flattening/boolbv_quantifier.cpp (hunk 4 above)
Comment (NlightNFotis) [reply src/solvers/flattening/boolbv_quantifier.cpp]: Clarified the comment.

File: src/util/simplify_expr_boolean.cpp (hunk 3 above)
Comment (NlightNFotis) [reply src/util/simplify_expr_boolean.cpp]: Yeah, you're right, I've changed to just handle take the info from `new_operands[0]` without any further processing.

File: src/util/simplify_expr_boolean.cpp (hunk 2 above)
Comment (NlightNFotis) [reply src/util/simplify_expr_boolean.cpp]: Done.

**Review by esteffin — APPROVED** [review]

**Comment by remi-delmas-3000** [comment]
@NlightNFotis I ran more tests with this fix and did not observe negative effects, approving

**Comment by NlightNFotis** [comment]
Cheers @remi-delmas-3000 , thanks for having a look.
